eyeshade/workers/reports.js

```diff
@@ -815,6 +816,153 @@ const date2objectId = (iso8601, ceilP) => {
+        amount: { $sum: '$amount' },
+        probi: { $sum: '$probi' },
+        fees: { $sum: '$fees' },
+        commission: { $sum: '$commission' }
+      }
+    }
+  ])
+  referralSettlementSummaries.forEach((summary) => {
+    summary.publisher = summary._id.publisher
+    summary.currency = summary._id.currency
+    summary.probi = new BigNumber(summary.probi)
+
+    if (!statements[summary.publisher]) statements[summary.publisher] = statementTemplate
+    statements[summary.publisher].settlements.summaries.push(summary)
+  })
+
+  underscore.keys(statements).forEach((publisher) => {
+    let pubOwner = ''
+    let balance = new BigNumber(0)
+    if (statements[publisher].referrals.summary.probi) {
+      balance = balance.plus(statements[publisher].referrals.summary.probi)
+      pubOwner = statements[publisher]
+    }
+    statements[publisher].settlements.summaries.forEach((summary) => {
+      balance = balance.minus(summary.probi)
+    })
+    if (balance.lessThan(0)) {
+      throw new Error('Publisher has been overpaid')
+    }
+    statements[publisher].balance = {
+      publisher: publisher,
+      owner: pubOwner,
+      altcurrency: altcurrency,
+      probi: balance
+    }
+  })
+
+  return statements
+}
+
+const findEligPublishers = async (debug, runtime, publishers) => {
+  const publishersC = runtime.database.get('publishers', debug)
+
+  if (!publishers) {
+    return []
+  }
+
+  const entries = await publishersC.find({ $or: publishers.map((pub) => { return { publisher: pub } }),
+    authorized: true,
+    verified: true }, { publisher: 1 })
+  return underscore.map(entries, (entry) => { return entry.publisher })
+}
+
+/**
+ * Prepare a json datastructure consisting of transactions to pay out the
+ * current balance owed to eligible publisher from referrals in the format expected
+ * by our payment tooling
+ **/
+const prepareReferralPayout = async (debug, runtime, authority, reportId, thresholdProbi) => {
+  const owners = runtime.database.get('owners', debug)
+
+  const statements = await referralStatement(debug, runtime, undefined, true)
+  const threshPubs = underscore.filter(underscore.keys(statements), (publisher) => {
+    return statements[publisher].balance.probi.greaterThan(thresholdProbi)
+  })
+  const eligPublishers = await findEligPublishers(debug, runtime, threshPubs)
+
+  const payments = []
+  for (let i = 0; i < eligPublishers.length; i++) {
+    const payment = statements[eligPublishers[i]].balance
+    payment.fees = payment.probi.times(feePercent)
+    payment.probi = payment.probi.minus(payment.fees)
+    payment.authority = authority
+    payment.transactionId = reportId
+
+    const entry = await owners.findOne({ owner: payment.owner })
+    if ((!entry) || (!entry.provider) || (!entry.parameters)) {
+      await notification(debug, runtime, entry.owner, payment.publisher, { type: 'verified_no_wallet' })
+      continue
+    }
+
+    const wallet = await runtime.wallet.status(entry)
```

Comment:
the call to `runtime.wallet.status` should be in a `try-catch` in case the underlying call to the uphold API fails... e.g.,

        } catch (ex) {
            await notification(debug, runtime, entry.owner, payment.publisher, { type: 'verified_invalid_wallet' })
        }

Reply:
:+1: